deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/networking/SilentTrainingDriver.java

```diff
@@ -0,0 +1,207 @@
+            //accumulator.receiveUpdate(message.getUpdates());
+        } else if (params != null && stepFunction != null) {
+
+            /*
+                this condition is only possible in spark local mode.
+                and it means that updates were already applied as local updates
+             */
+            if (transport.numberOfKnownClients() == 1) {
+                return;
+            }
+
+            // master invokes everything, since that's Silent Worker approach: we want master to be always up-to-date
+            synchronized (this) {
+                // threshold decoder is inplace & fast
+                Nd4j.getExecutioner().thresholdDecode(message.getUpdates(), updates);
+
+                // this simple flag shows that we have something not applied, will be used at finishTraining() method
+                hasSomething.set(true);
+
+                // we apply updates every X iterations, and we don't really need X to be small here
+                if (updatesCount.incrementAndGet() % Math.max(transport.numberOfKnownClients(), 5) == 0) {
+                    stepFunction.step(params, updates);
+
+                    // once accumulated updates are applied - reset storage, and wait for other messsages
+                    Nd4j.getMemoryManager().memset(updates);
+                    hasSomething.set(false);
+                }
+            }
+
+            // we should echo this message to everyone but this shard, but only if there's > 1 shard/client available
+            if (transport.numberOfKnownClients() > 1) {
+                //log.info("Resending message, skipping {}", message.getOriginatorId());
+                transport.sendMessageToAllClients(message, message.getOriginatorId(), transport.getOwnOriginatorId());
+            } // else log.info("No known Clients so far");
+        } else
+            throw new DL4JInvalidConfigException("Neither GradientsAccumulator or StepFunction is defined!");
+    }
+
+    @Override
+    public void pickTraining(SilentUpdatesMessage message) {
+        // this message won't be ever called
+    }
+
+    @Override
+    public void aggregationFinished(VoidAggregation aggregation) {
+        // no-op
+    }
+
+    /**
+     * This method is used on Master only, applies buffered updates to params
+     *
+     * @param originatorId
+     * @param taskId
+     */
+    @Override
+    public void finishTraining(long originatorId, long taskId) {
+        // on Master thread we'll be applying final gradients
+        if (params != null && stepFunction != null) {
+            if (hasSomething.get()) {
+                stepFunction.step(params, updates);
+                Nd4j.getMemoryManager().memset(updates);
+            }
+        }
+    }
+
+    @Override
+    public void addCompletionHook(long originatorId, long frameId, long messageId) {
+        // no-op
+    }
+
+    @Override
+    public String targetMessageClass() {
+        return SilentUpdatesMessage.class.getSimpleName();
```

Comment:
That's for provider. Simple name works fine, since provide does mapping. However, not used too.